### PR TITLE
Add ContainerSecurityContext to Pools

### PR DESCRIPTION
**Pull request by allanrogerr**
Modified tenant CRD to support CSC under pools
Modified tenant statefulset to support CSC
Added support for CSC on Helm chart
Added implementation examples to 
- examples/kustomization/base/tenant.yaml
- examples/kustomization/tenant-lite/tenant.yaml
Avoided operator upgrade tests where example yamls do not correspond to the operator binary. In this case, the test chooses a lower version as being the current branch of the operator

Todo
Make sure that if the kubernetes version doesn't support CSC to stop the sync of the tenant with an error (tenant.Validate()) in theory <1.22

### CODE REVIEW

**Review by dvaldivia — APPROVED** [review]
I tested all the way down to k8s `v1.19` and it seems to be working fine.

**Comment by cniackz** [comment]
Tested locally, couple of comments:

1. We already put the securityContext of a container prior this change, tested but we couldn't modify it.
2. With this change, we can now modify it, which is good and tested.
3. But I still see missing fields like `fsGroupChangePolicy` that if set not to always it can shorten the time to deploy our pod.

Please read at:
https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods

Specifically at:
```
OnRootMismatch: Only change permissions and ownership if the permission and the ownership of root directory does not match with expected permissions of the volume. This could help shorten the time it takes to change ownership and permission of a volume.
```

So I suggest to apply any missing field on `securityContext` to provide more configuration options, all possible is the ideal.

**Review by cniackz — CHANGES_REQUESTED** [review]
Please add missing fields on `securityContext` like: `fsGroupChangePolicy`

**Comment by cniackz** [comment]
But overall good Job Allan, you impressed me on how fast you got this change to properly work in k8s 👍

**Comment by dvaldivia** [comment]
there's no `fsGroupChangePolicy` for Container SecurityContext it seems

**Comment by allanrogerr** [comment]
> Tested locally, couple of comments:
> 
> 1. We already put the securityContext of a container prior this change, tested but we couldn't modify it.
> 2. With this change, we can now modify it, which is good and tested.
> 3. But I still see missing fields like `fsGroupChangePolicy` that if set not to always it can shorten the time to deploy our pod.
> 
> Please read at: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods
> 
> Specifically at:
> 
> ```
> OnRootMismatch: Only change permissions and ownership if the permission and the ownership of root directory does not match with expected permissions of the volume. This could help shorten the time it takes to change ownership and permission of a volume.
> ```
> 
> So I suggest to apply any missing field on `securityContext` to provide more configuration options, all possible is the ideal.

Thanks @dvaldivia - @cniackz When reviewing ContainerSecurityContext (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#securitycontext-v1-core), I noted that the available fields do not include several from PodSecurityContext (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podsecuritycontext-v1-core).

**Comment by allanrogerr** [comment]
> I tested all the way down to k8s `v1.19` and it seems to be working fine.

@dvaldivia Do we still need the floor test then? It may be useful in the future. Let me know.

**Comment by dvaldivia** [comment]
yes, we merged https://github.com/minio/operator/pull/1304 but we set the floor to `v1.21` of k8s @allanrogerr

**Review by cniackz — APPROVED** [review]
LGTM

**Review by dilverse — DISMISSED** [review]
Awesome stuff @allanrogerr! 🤘

**Comment by allanrogerr** [comment]
fixes: https://github.com/minio/operator/issues/1366

**Comment by allanrogerr** [comment]
fixes: https://github.com/minio/operator/issues/1310

**Comment by allanrogerr** [comment]
fixes: https://github.com/minio/operator/issues/1267

**Comment by allanrogerr** [comment]
@dvaldivia @cniackz I have added the defaulting capability as requested
https://github.com/minio/operator/pull/1372/commits/0bd5ab79c2e9834771286512207d02cda1925729

**Comment by allanrogerr** [comment]
Tests
1.- Pull allanrogerr:add-containersecuritycontext-helm
2.- After commenting out "destroy_kind", run ./testing/check-helm.sh 
3.- Observe operator and console without Container Security Context
4.- After commenting out "destroy_kind", run ./testing/deploy-tenant.sh 
5.- Observe operator and console without Container Security Context

6.- Modify temporarily
	helm/operator/values.yaml
		Add the following yaml snippet under "console" and "operator"
	helm/tenant/values.yaml
		Add the following yaml snippet under "pools"

		containerSecurityContext:
	    runAsUser: 1000
	    runAsGroup: 1000
	    runAsNonRoot: true
	    allowPrivilegeEscalation: false
	    readOnlyRootFilesystem: false
	    capabilities:
	      add:
	        - ALL
	      drop:
	        - ALL

	examples/kustomization/tenant-lite/tenant.yaml
		Add the following yaml snippet under "pools"

		containerSecurityContext:
	    runAsUser: 1000
	    runAsGroup: 1000
	    runAsNonRoot: true
	    allowPrivilegeEscalation: false
	    readOnlyRootFilesystem: false
	    capabilities:
	      add:
	        - ALL
	      drop:
	        - ALL

	resources/base/console-ui.yaml
	resources/base/deployment.yaml
		Add the following yaml snippet under "containers"

		securityContext:
      runAsUser: 1000
      runAsGroup: 1000
      runAsNonRoot: true
      allowPrivilegeEscalation: false
      readOnlyRootFilesystem: false
      capabilities:
        add:
          - ALL
        drop:
          - ALL

7.- Run ./testing/check-helm.sh again
8.- After the test completes, observe operator, console and tenant output yamls with the above Container Security Context
	e.g.
		k get pod/minio1-pool-0-0 -n default -o yaml
		k get pod/console-68766cb74-x5hwn -n minio-operator -o yaml
		k get pod/minio-operator-8bbdb59d6-94j4m -n minio-operator -o yaml

9.- Run ./testing/deploy-tenant.sh again
10.- After the test completes, observe operator, console and tenant output yamls with the above Container Security Contexts
	e.g.
		 k get pod/storage-lite-pool-0-0 -n tenant-lite -o yaml
		 k get pod/console-7f7699948f-hjrq7 -n minio-operator -o yaml
		 k get pod/minio-operator-95c4dc86-zs5s9 -n minio-operator -o yaml

**Review by dvaldivia — DISMISSED** [review]

**Comment by cniackz** [comment]
Read comments, and review the code changes, and it looks good to me.
Last time I tested this PR with Kustomize, this time I will test is with Helm and if all goes well, then I will approve and notify Allan about it.

File: helm/tenant/values.yaml [hunk 1]
```diff
@@ -61,6 +61,15 @@ tenant:
       topologySpreadConstraints: [ ]
       ## Configure Runtime Class
       # runtimeClassName: ""
+      securityContext:
```
Comment (cniackz) [inline helm/tenant/values.yaml:64]: `securityContext` is already defined above as empty braces:

```yaml
      ## Configure security context
      securityContext: { }
```

Shouldn't we keep only one copy of this value?

**Review by cniackz — COMMENTED** [review]
Allan, minor comment for `helm/tenant/values.yaml`, please take a look and keep one copy of `securityContext` in the values.

**Review by dilverse — APPROVED** [review]

**Comment by cniackz** [comment]
Thank you for the fix Allan :)

**Review by cniackz — APPROVED** [review]
LGTM

**Review by pjuarezd — APPROVED** [review]

**Pull request merged**